pkg/workloads/console/controller.go

```diff
@@ -206,6 +229,75 @@ func (r *reconciler) updateStatus(job *batchv1.Job) error {
 	return nil
 }
 
+func calculateStatus(csl *workloadsv1alpha1.Console, job *batchv1.Job, pod *corev1.Pod) workloadsv1alpha1.ConsoleStatus {
+	newStatus := csl.DeepCopy().Status
+
+	// We may have been passed an empty Job struct, if the job no longer exists,
+	// so determine whether it's a real job by checking if it has a name.
+	if job != nil && len(job.Name) != 0 {
+		// We want to give the console session *at least* the time specified in the
+		// timeout, therefore base the expiry time on the job creation time, rather
+		// than the console creation time, to take into account any delays in
+		// reconciling the console object.
+		// TODO: We may actually want to use a base of when the Pod entered the
+		// Running phase, as image pull time could be significant in some cases.
+		jobCreationTime := job.ObjectMeta.CreationTimestamp.Time
+		expiryTime := metav1.NewTime(
+			jobCreationTime.Add(time.Second * time.Duration(csl.Spec.TimeoutSeconds)),
+		)
+		newStatus.ExpiryTime = &expiryTime
+		if pod != nil {
+			newStatus.PodName = pod.ObjectMeta.Name
+		}
+
+		newStatus.Phase = calculatePhase(job, pod)
+	}
+
+	return newStatus
+}
+
+// Get a job's current status
+func calculatePhase(job *batchv1.Job, pod *corev1.Pod) workloadsv1alpha1.ConsolePhase {
+	// Currently a job can only have two conditions: Complete and Failed
+	// Both indicate that the console has stopped
+	for _, c := range job.Status.Conditions {
+		if c.Type == batchv1.JobComplete || c.Type == batchv1.JobFailed {
+			return workloadsv1alpha1.ConsoleStopped
+		}
+	}
+
+	// If the pod exists and is running, then the console is running
+	if pod != nil && pod.Status.Phase == corev1.PodRunning {
+		return workloadsv1alpha1.ConsoleRunning
+	}
+
+	// Otherwise, assume the console is pending (i.e. still starting up)
+	return workloadsv1alpha1.ConsolePending
+}
+
+func requeueForExpiration(logger kitlog.Logger, status workloadsv1alpha1.ConsoleStatus) k8rec.Result {
```

Comment:
Added this back from the previous version because we want to update the status when the job expires. We don't (currently) watch for events on our child resources so we won't otherwise hear about this.